numpy/random/bit_generator.pyx

```diff
@@ -0,0 +1,458 @@
+        if out_dtype == np.dtype(np.uint32):
+            pass
+        elif out_dtype == np.dtype(np.uint64):
+            n_words *= 2
+        else:
+            raise ValueError("only support uint32 or uint64")
+        state = np.zeros(n_words, dtype=np.uint32)
+        src_cycle = cycle(self.pool)
+        for i_dst in range(n_words):
+            data_val = next(src_cycle)
+            data_val ^= hash_const
+            hash_const *= MULT_B
+            data_val *= hash_const
+            data_val ^= data_val >> XSHIFT
+            state[i_dst] = data_val & MASK32
+        if out_dtype == np.dtype(np.uint64):
+            state = state.view(np.uint64)
+        return state
+
+    def spawn(self, n_children):
+        """ Spawn a number of child `SeedSequence`s by extending the
+        `spawn_key`.
+        Parameters
+        ----------
+        n_children : int
+        Returns
+        -------
+        seqs : list of `SeedSequence`s
+        """
+        seqs = []
+        for i in range(self.n_children_spawned,
+                       self.n_children_spawned + n_children):
+            seqs.append(SeedSequence(
+                self.entropy,
+                program_entropy=self.program_entropy,
+                spawn_key=self.spawn_key + (i,),
+                pool_size=self.pool_size,
+            ))
+        return seqs
+
+cdef class BitGenerator():
+    """
+    BitGenerator(seed=None)
+
+    Abstract Base Class for generic BitGenerators, which provide a stream
+    of random bits based on different algorithms.
+
+    Parameters
+    ----------
+    seed: {None, SeedSequence, int, sequence[int]}, optional
+        A SeedSequence to initialize the BitGenerator. If None, one will be
+        created. If an int or a sequence of ints, it will be used as the
+        entropy for creating a SeedSequence.
+
+    Attributes
+    ----------
+    lock: threading.Lock
+        Lock instance that is shared so that the same bit git generator can
+        be used in multiple Generators without corrupting the state. Code that
+        generates values from a bit generator should hold the bit generator's
+        lock.
+
+    SeeAlso
+    -------
+    `SeedSequence`
+    """
+
+    def __init__(self, seed=None):
+        self.lock = Lock()
+
+        self._ctypes = None
+        self._cffi = None
+
+        cdef const char *name = "BitGenerator"
+        self.capsule = PyCapsule_New(<void *>&self._bitgen, name, NULL)
+        if seed is None:
+            seed = SeedSequence()
+        elif isinstance(seed, SeedSequence):
+            pass
+        elif isinstance(seed, (int, np.integer)):
+            seed = SeedSequence(seed)
+        elif isinstance(seed, (list, tuple, np.ndarray)):
+            seed = SeedSequence(seed)
+        else:
+            raise TypeError('{}(seed) expects int or SeedSequence for seed '
+                             'not {}'.format(type(self).__name__, seed))
+        self._seed = seed
```

Comment:
How does the seed get set into the bit generators state?  In the old model, you would not call `self.seed(self._seed)`

Reply:
One could use a private class self._set_seed(seed) that is requried by children that handles specific seeding.

Reply:
https://github.com/numpy/numpy/pull/13780/files/82453cdb6cb2d0d99b2394a78e1af13083324f30#diff-81acdee7a38ef6ded447175da539d1a6R105

Reply:
I don't think I understand.  I was thinking that the the `BitGenerator` class would call `self._set_seed(self._seed)` (which is provided by `PCG64._set_seed(self, value)`) at the end of  its `__init__`.  Might not be worth the hassle since then `PCG64.__init__` would have to ensure that everything existed before the super call (which is fine for PCG64, but not for a generator that needs aligned memory).

Reply:
The BitGenerators call `SeedSequence.generate_state`, then use that value to set their seed. in particular PCG64 calls 
```
val = self._seed_seq.generate_state(4, np.uint64); 
pcg64_set_seed(&self.rng_state,
                       <uint64_t *>np.PyArray_DATA(val),
                       <uint64_t *>np.PyArray_DATA(val) + 2)
```